### PR TITLE
Migrate type validation to Bun, enhance API client and improve React keys

### PR DESCRIPTION
**Pull request by WilliamAGH**
This pull request introduces a major change to how globally unique type/interface/enum names are enforced in the codebase, replacing a nondeterministic ESLint rule with a deterministic build-time script. It also improves validation and error handling in the OpenAI-compatible client and message-mapper, enhances test environment robustness, and includes some minor UI and code improvements.

**Type Uniqueness Enforcement**

* Removes the custom ESLint rule `project/no-duplicate-types` from `eslint.config.ts` and replaces it with a new deterministic build-time script, `scripts/check-duplicate-types.ts`, which scans all type files and reports duplicates with consistent output. This avoids nondeterministic warnings caused by ESLint's file-processing order. [[1]](diffhunk://#diff-3202509a51dddde1968b20a04dfcedc90615eb6681cca0cda5f43a418a516857L8-R8) [[2]](diffhunk://#diff-3202509a51dddde1968b20a04dfcedc90615eb6681cca0cda5f43a418a516857L19-L20) [[3]](diffhunk://#diff-3202509a51dddde1968b20a04dfcedc90615eb6681cca0cda5f43a418a516857L230-R228) [[4]](diffhunk://#diff-74bc8b89e7837a2eb333cc3a0df70406b60b9ee3ddd6d7c000cd45b46dd0e1f9R1-R99)

**OpenAI-Compatible Client and Message Mapper Improvements**

* Adds request validation for the responses endpoint using a new Zod schema (`openAiCompatibleResponsesRequestSchema`), ensuring stricter input validation before making API calls. [[1]](diffhunk://#diff-9d02cb4b2c5dcdd22c40b452588846cf485037f45dae1b8ad78e6d016557b424R13) [[2]](diffhunk://#diff-9d02cb4b2c5dcdd22c40b452588846cf485037f45dae1b8ad78e6d016557b424R76-R86) [[3]](diffhunk://#diff-9d02cb4b2c5dcdd22c40b452588846cf485037f45dae1b8ad78e6d016557b424R216-R221) [[4]](diffhunk://#diff-9d02cb4b2c5dcdd22c40b452588846cf485037f45dae1b8ad78e6d016557b424R241-R246) [[5]](diffhunk://#diff-2ce092b74fddadcce909f029fae1dcdf1b2d069dc0bee476d5a2c1967c584f7bR128-R160)
* Improves timeout handling in `resolveClient` by logging a warning and defaulting to a safe value if no timeout is provided.
* Enhances error messages and model handling in streaming responses and model name normalization, including better fallbacks and warnings for missing or malformed data. [[1]](diffhunk://#diff-9d02cb4b2c5dcdd22c40b452588846cf485037f45dae1b8ad78e6d016557b424R285-R290) [[2]](diffhunk://#diff-d004bf30e1bbe755fe3ea256a8ac0cfc9bf9fcdf2f36c7fe9b28d4cc94145e0aL20-R28)
* Improves tool parameter handling in message mapping, logging warnings and defaulting to empty schemas when parameters are missing.
* Adds support for assistant message refusals in `toResponsesInput`, ensuring refusal content is preserved and correctly mapped. [[1]](diffhunk://#diff-d004bf30e1bbe755fe3ea256a8ac0cfc9bf9fcdf2f36c7fe9b28d4cc94145e0aR139-R146) [[2]](diffhunk://#diff-d004bf30e1bbe755fe3ea256a8ac0cfc9bf9fcdf2f36c7fe9b28d4cc94145e0aL134-R169) [[3]](diffhunk://#diff-d004bf30e1bbe755fe3ea256a8ac0cfc9bf9fcdf2f36c7fe9b28d4cc94145e0aR185-R191)

**Test Environment Robustness**

* Makes test environment URL detection more robust by using `Reflect.get` to safely access `globalThis.location` properties, avoiding potential runtime errors.

**Minor UI and Code Improvements**

* Updates a link in the investments client component to point to `/contact` instead of `/`.
* Fixes a React key warning in the terminal history component by generating more unique keys for chat output lines.

---

> [!NOTE]
> **Medium Risk**
> Touches the AI chat streaming/error contract end-to-end (new error kinds, status mapping, and sanitized messages) plus upstream client validation/timeout behavior, which could affect production chat reliability if mismatched across layers.
> 
> **Overview**
> This PR replaces the custom ESLint cross-file type-uniqueness rule with a deterministic Bun script (`scripts/check-duplicate-types.ts`) and wires it into `lint:checks`/`validate:build`, removing the old ESLint plugin and updating docs accordingly.
> 
> It also introduces semantic AI chat stream error classification (`kind`) including timeout detection/mapping to `504`, propagates that field through SSE (`sse-stream.ts`) to the browser client and terminal UI, and adjusts user-facing errors to avoid leaking upstream details. Separately, it tightens OpenAI-compatible Responses request validation (new Zod schema + tool normalization), reduces default upstream timeout/retries, improves message mapping for tool parameters/refusals, and includes small UX/test/data tweaks (terminal welcome message factory + Enter key test, investments link to `/contact`, projects ordering/date, safer test base URL lookup, Sentry `skipOpenTelemetrySetup`).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 24ad92e28ec0c83d881da465578126911ced5764. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

<h2>Greptile Overview</h2>

<h3>Greptile Summary</h3>

This PR successfully migrates type uniqueness enforcement from a nondeterministic ESLint rule to a deterministic build-time Bun script, and significantly hardens the OpenAI-compatible AI client with proper validation, error classification, and refusal handling.

**Major Changes:**

- **Type Uniqueness Enforcement**: Replaces `eslint/no-duplicate-types` rule with `scripts/check-duplicate-types.ts`, wired into `lint:checks` and `validate:build`. The new script provides deterministic, sorted output and includes fallback base-ref resolution for CI environments.

- **Responses API Validation**: Introduces comprehensive Zod validation for Responses requests (`openAiCompatibleResponsesRequestSchema`) with proper tool normalization (parameters and strict fields default to `null` per SDK contract). Request validation now happens before upstream calls, preventing malformed payloads.

- **Error Classification**: Adds semantic error kinds (`timeout`, `auth`, `rate_limit`, `model_unavailable`, `upstream`) with explicit timeout detection returning 504 status. Upstream auth failures (401/403) and rate limits (429) are remapped to 503 to avoid leaking topology, with comprehensive test coverage.

- **Refusal Handling**: Assistant messages in Responses input now properly support the `refusal` field as a fallback when `content` is missing, with `Reflect.get()` used to safely extract the field. Tool-call messages correctly emit assistant content or refusal before function calls, with `continue` preventing duplication.

- **Robustness Improvements**: Timeout handling warns for invalid values (NaN/Infinity) and defaults to 30s. Tool parameters missing in Chat Completions default to empty schema with warnings. Model name parsing handles empty segments gracefully. Default max retries reduced from 3 to 1, and timeout from 60s to 30s for faster failure detection.

- **UI/UX Fixes**: Terminal welcome message converted to factory function ensuring unique timestamps on resets. React key warning resolved in terminal history. Environment URL detection robustified with `Reflect.get()`.

All previous thread issues have been addressed with documented commit references.

<h3>Confidence Score: 4/5</h3>

- Safe to merge with minor observations to verify in production
- The PR includes thorough validation improvements and comprehensive test coverage for error handling. Type uniqueness enforcement is now deterministic and wired into CI. The main risk areas are the reduced timeout/retry defaults (30s/1 retry vs 60s/3 retries) and the addition of request validation that could potentially reject valid edge-case payloads. However, the validation schemas are carefully aligned with OpenAI SDK contracts, previous issues were all addressed, and the error classification provides better observability. The changes are well-tested and follow established patterns.
- Monitor `openai-compatible-client.ts` and `openai-compatible-message-mapper.ts` in production for any validation rejections or timeout behavior changes due to reduced defaults

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| scripts/check-duplicate-types.ts | New deterministic build-time script replaces ESLint rule for type uniqueness checks. Includes proper fallback base-ref resolution and sorted output for consistent results. |
| package.json | Integrates `check:duplicate-types` into lint and build validation pipelines. Makes ESLint CI non-blocking and removes redundant script aliases. |
| src/lib/ai/openai-compatible/openai-compatible-client.ts | Adds Responses request validation via Zod, improves timeout handling with warnings for invalid values, normalizes tool parameters, and enhances error messages. Reduces default timeout to 30s and max retries to 1. |
| src/lib/ai/openai-compatible/openai-compatible-message-mapper.ts | Adds refusal handling for assistant messages in Responses input mapping, improves tool parameter validation with fallback to empty schema, and strengthens model name parsing with warnings. |
| src/types/schemas/ai-openai-compatible.ts | Introduces dedicated Responses request/input schemas with proper refusal support, tool choice variants, and nullable field handling aligned with OpenAI SDK contracts. |
| src/app/api/ai/chat/[feature]/upstream-error.ts | Adds semantic error kind classification (timeout, auth, rate_limit, model_unavailable, upstream) with explicit timeout detection returning 504 status. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Browser
    participant SSE as SSE Stream Handler
    participant Client as OpenAI Client
    participant Validator as Request Validator
    participant Mapper as Message Mapper
    participant Upstream as AI Service

    Browser->>SSE: POST /api/ai/chat/[feature]
    SSE->>Validator: validateResponsesRequest(request)
    Validator->>Validator: Parse with Zod schema
    Validator->>Validator: Normalize tools (parameters ?? null, strict ?? null)
    Validator-->>SSE: validatedRequest

    SSE->>Mapper: toResponsesInput(messages)
    Mapper->>Mapper: Extract refusal from assistant messages
    Mapper->>Mapper: Map tool_calls to function_call items
    Mapper->>Mapper: Fallback content to refusal if empty
    Mapper-->>SSE: ResponseInput[]

    SSE->>Client: client.responses.stream(validatedRequest)
    Client->>Upstream: HTTP POST with validated payload
    
    alt Success Path
        Upstream-->>Client: Stream events
        Client-->>SSE: message_start event
        SSE-->>Browser: SSE: message_start
        loop Stream deltas
            Upstream-->>Client: output_text.delta
            Client-->>SSE: onDelta callback
            SSE-->>Browser: SSE: message_delta
        end
        Client-->>SSE: finalResponse
        SSE->>SSE: Normalize output_text
        SSE-->>Browser: SSE: message_done
    else Timeout Error
        Upstream-->>Client: Timeout/no response
        Client-->>SSE: Error thrown
        SSE->>SSE: isTimeoutError() → true
        SSE->>SSE: resolveErrorResponse() → {status: 504, kind: "timeout"}
        SSE-->>Browser: SSE: error {kind: "timeout"}
    else Auth/Rate Limit Error
        Upstream-->>Client: 401/403/429
        Client-->>SSE: Error with status
        SSE->>SSE: resolveErrorResponse() → {status: 503, kind: "auth"|"rate_limit"}
        SSE-->>Browser: SSE: error (sanitized, no upstream details)
    else Model Unavailable
        Upstream-->>Client: 400 "Failed to load model"
        Client-->>SSE: Error with message
        SSE->>SSE: isModelLoadFailure() → true
        SSE->>SSE: resolveErrorResponse() → {status: 503, kind: "model_unavailable"}
        SSE-->>Browser: SSE: error {kind: "model_unavailable"}
    else Generic Upstream Error
        Upstream-->>Client: Other error
        Client-->>SSE: Error
        SSE->>SSE: resolveErrorResponse() → {status: 502, kind: "upstream"}
        SSE-->>Browser: SSE: error {kind: "upstream"}
    end

    Browser->>Browser: Parse error kind for UI display
```

**Context used:**

- Context from `dashboard` - CLAUDE.md ([source](https://app.greptile.com/review/custom-context?memory=23822fbb-4a3d-480a-90b4-1b28f83b089c))
- Context from `dashboard` - AGENTS.md ([source](https://app.greptile.com/review/custom-context?memory=936bf778-c6d3-49e7-b054-8739a9e6d9ed))

### CODE REVIEW

File: eslint.config.ts [hunk 1]
```diff
@@ -227,21 +224,8 @@ const config = defineConfig(
...
-          "no-duplicate-types": noDuplicateTypesRule,
-        },
-      },
-    },
-    rules: {
-      "project/no-duplicate-types": "warn", // Changed to warn for gradual migration
-    },
-  },
+  // Project-specific global type uniqueness is enforced by `bun scripts/check-duplicate-types.ts`
+  // (deterministic build-time check, not an ESLint rule — see scripts/check-duplicate-types.ts)
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply eslint.config.ts]: # Suggestion Implemented ✅
## Changes Made
- Wired duplicate-type enforcement into standard validation paths.
- Added \ and included it in \ and \.

## Validation
- \Found 0 warnings and 0 errors.
Finished in 345ms on 897 files with 255 rules using 16 threads.
[check:duplicate-types] No duplicate type names found.
🔍 Scanning TypeScript files...
📂 Found 744 TypeScript files
🔗 Detecting circular dependencies...
⚠️  Found 0 circular dependencies
✅ No circular dependencies found! passed with 0 lint/type errors.

## Commit
- \ fix(validate): enforce duplicate type checks in lint and build validation

File: scripts/check-duplicate-types.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+  if (duplicateCount > 0) {
+    console.error(
+      `\n[check:duplicate-types] ${duplicateCount} duplicate type name(s) found. All type names must be globally unique.\n`,
+    );
+    process.exit(1);
+  }
+
+  console.log("[check:duplicate-types] No duplicate type names found.");
+}
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply scripts/check-duplicate-types.ts]: # Suggestion Implemented ✅
## Changes Made
- Integrated \[check:duplicate-types] No duplicate type names found. into the main validation path so CI/dev validation runs it automatically.

## Validation
- \Found 0 warnings and 0 errors.
Finished in 359ms on 897 files with 255 rules using 16 threads.
[check:duplicate-types] No duplicate type names found.
🔍 Scanning TypeScript files...
📂 Found 744 TypeScript files
🔗 Detecting circular dependencies...
⚠️  Found 0 circular dependencies
✅ No circular dependencies found! passed with 0 lint/type errors.

## Commit
- \ fix(validate): enforce duplicate type checks in lint and build validation

File: src/lib/ai/openai-compatible/openai-compatible-client.ts [hunk 3]
```diff
@@ -10,6 +10,7 @@ import {
   type OpenAiCompatibleResponsesResponse,
   openAiCompatibleChatCompletionsRequestSchema,
   openAiCompatibleChatCompletionsResponseSchema,
+  openAiCompatibleResponsesRequestSchema,
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/lib/ai/openai-compatible/openai-compatible-client.ts]: # Suggestion Implemented ✅
## Changes Made
- Updated Responses request validation to return the parsed payload rather than the raw input.
- Added explicit normalization for OpenAI Responses tool fields (\/\) to satisfy SDK request typing.

## Validation
- \Found 0 warnings and 0 errors.
Finished in 340ms on 897 files with 255 rules using 16 threads.
[check:duplicate-types] No duplicate type names found.
🔍 Scanning TypeScript files...
📂 Found 744 TypeScript files
🔗 Detecting circular dependencies...
⚠️  Found 0 circular dependencies
✅ No circular dependencies found! passed with 0 lint/type errors.

## Commit
- \ fix(ai): return parsed responses request and harden timeout validation

[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/lib/ai/openai-compatible/openai-compatible-client.ts]: # Suggestion Implemented ✅
## Changes Made
- Updated Responses request validation to return the parsed payload instead of the raw input.
- Added explicit normalization for OpenAI Responses tool fields (`parameters` and `strict`) to satisfy SDK request typing.

## Validation
- `bun run validate` passed with 0 lint/type errors.

## Commit
- `d38fc7ff` fix(ai): return parsed responses request and harden timeout validation

File: src/types/schemas/ai-openai-compatible.ts [hunk 4]
```diff
@@ -125,6 +125,39 @@ export type OpenAiCompatibleChatCompletionsRequest = z.infer<
   typeof openAiCompatibleChatCompletionsRequestSchema
 >;
 
+const openAiCompatibleResponsesFunctionToolSchema = z.object({
+  type: z.literal("function"),
+  name: z.string().min(1),
+  description: z.string().nullable().optional(),
+  parameters: z.record(z.string(), z.unknown()).nullable().optional(),
+  strict: z.boolean().optional(),
+});
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/types/schemas/ai-openai-compatible.ts]: # Suggestion Analysis 🔍
## Decision: Not Implemented
## Reasoning
- I verified the installed dependency contract directly in this repo for OpenAI SDK `6.18.0`.
- `node_modules/openai/resources/responses/responses.d.ts:289-303` defines Responses `FunctionTool` with `parameters: { [key: string]: unknown } | null` and `strict: boolean | null`.
- This PR parses and normalizes requests before passing them through the SDK-typed request shape, so nullable tool fields are intentionally supported for contract alignment.
- I also implemented explicit normalization in the client path so missing tool fields are converted to `null` before send.

## Conclusion
Keeping nullable support here is intentional to match the installed SDK surface and avoid schema/runtime type drift.

File: src/lib/ai/openai-compatible/openai-compatible-message-mapper.ts [hunk 5]
```diff
@@ -151,6 +182,13 @@ export function toResponsesInput(
     }
 
     const role: EasyInputMessage["role"] = message.role;
+    if (message.role === "assistant") {
+      const refusalContent = getAssistantRefusal(message);
+      if (refusalContent) {
+        items.push({ type: "message", role, content: refusalContent });
+        continue;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/lib/ai/openai-compatible/openai-compatible-message-mapper.ts]: Suggestion Implemented ✅

Changes Made
- Updated assistant mapping fallback in src/lib/ai/openai-compatible/openai-compatible-message-mapper.ts:185 so refusal is only used when assistant content is missing/empty.
- Preserved normal assistant content when both content and refusal are present.

Validation
- bun run validate passed with 0 lint/type errors.

Commit
- bec14b9e fix(ai): keep assistant content over refusal text

File: scripts/check-duplicate-types.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    duplicateCount++;
+    console.error(`\n  Type "${name}" is declared ${locations.length} times:`);
+    for (const loc of locations) {
+      console.error(`    ${loc.file}:${loc.line}`);
+    }
+  }
+
+  return duplicateCount;
+}
+
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply scripts/check-duplicate-types.ts]: Suggestion Implemented ✅

Changes Made
- Added fallback base-ref detection in scripts/check-duplicate-types.ts.
- Script now resolves the first available ref from origin/main, origin/dev, main, dev.
- If no fallback ref exists, it logs and checks all declarations explicitly.

Validation
- bun run validate passed with 0 lint/type errors.

Commit
- e46dd213 fix(lint): detect fallback base ref for duplicate type checks

File: src/lib/ai/openai-compatible/openai-compatible-client.ts [hunk 7]
```diff
@@ -63,6 +73,36 @@ function validateChatRequest(request: OpenAiCompatibleChatCompletionsRequest) {
...
+      ...tool,
+      parameters: tool.parameters ?? null,
+      strict: tool.strict ?? null,
+    }),
+  );
+
+  return {
+    ...request,
+    model: parsedRequest.model,
+    input: parsedRequest.input,
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/lib/ai/openai-compatible/openai-compatible-client.ts]: Suggestion Implemented ✅

Changes Made
- Responses input validation now uses an assistant schema that accepts refusal for responses-mode input.
- Preserved refusal through validated request input so toResponsesInput() can read it.
- Responses mapper typing and client request typing were aligned with the responses input schema.

Validation
- bun run validate passed with 0 lint/type errors.

Commits
- 3bb1dd87 fix(ai): keep assistant refusal in responses input schema
- f7f62d16 fix(ai): type responses mapper refusal extraction safely
- 9ea42877 fix(ai): align responses client input typing with schema

File: src/types/schemas/ai-openai-compatible.ts [hunk 8]
```diff
@@ -125,6 +125,39 @@ export type OpenAiCompatibleChatCompletionsRequest = z.infer<
   typeof openAiCompatibleChatCompletionsRequestSchema
 >;
 
+const openAiCompatibleResponsesFunctionToolSchema = z.object({
+  type: z.literal("function"),
+  name: z.string().min(1),
+  description: z.string().nullable().optional(),
+  parameters: z.record(z.string(), z.unknown()).nullable().optional(),
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/types/schemas/ai-openai-compatible.ts]: Suggestion Analysis 🔍

Decision: Not Implemented

Reasoning
- OpenAI Responses FunctionTool expects parameters as a JSON object map (`{ [key: string]: unknown } | null`) in node_modules/openai/resources/responses/responses.d.ts.
- Current schema `z.record(z.string(), z.unknown()).nullable().optional()` matches that contract.
- `z.record(...)` already accepts standard JSON Schema shapes, including keys like `$schema`, nested objects, arrays, booleans, and numbers.
- Widening to `z.unknown()` would weaken boundary validation by allowing invalid non-object values.

Conclusion
- Keeping the current parameters schema preserves type-safe validation and SDK compatibility.

File: src/lib/ai/openai-compatible/openai-compatible-client.ts [hunk 9]
```diff
@@ -63,6 +73,36 @@ function validateChatRequest(request: OpenAiCompatibleChatCompletionsRequest) {
...
+    temperature: parsedRequest.temperature,
+    top_p: parsedRequest.top_p,
+    max_output_tokens: parsedRequest.max_output_tokens,
+    tools: normalizedTools,
+    tool_choice: request.tool_choice,
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/lib/ai/openai-compatible/openai-compatible-client.ts]: Suggestion Analysis 🔍

Decision: Not Implemented

Reasoning
- The request is already schema-validated (`openAiCompatibleResponsesRequestSchema.parse(request)`), so `tool_choice` is validated before use.
- Returning `request.tool_choice` preserves the stricter OpenAI SDK type from `ResponseCreateParamsNonStreaming` at compile time.
- Using `parsedRequest.tool_choice` directly would currently widen the static type because the schema branch allows generic object forms for compatibility, which is less precise than the SDK contract.

Conclusion
- Keeping `request.tool_choice` is intentional to preserve strict SDK typing while still validating at runtime.

File: src/lib/ai/openai-compatible/openai-compatible-client.ts [hunk 10]
```diff
@@ -63,6 +74,36 @@ function validateChatRequest(request: OpenAiCompatibleChatCompletionsRequest) {
...
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/lib/ai/openai-compatible/openai-compatible-client.ts]: Suggestion Analysis 🔍

Decision: Not Implemented

Reasoning
- `validateResponsesRequest()` still validates `tool_choice` via `openAiCompatibleResponsesRequestSchema.parse(request)` before returning.
- Using `parsedRequest.tool_choice` here widens static typing to include the schema record branch, which does not match the stricter OpenAI SDK `ResponseCreateParamsNonStreaming` tool-choice union.
- Keeping `request.tool_choice` preserves the SDK-level type contract at compile time while still enforcing runtime validation through the parse call.

Conclusion
- Current line is intentional for type safety + validated runtime behavior.

[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/lib/ai/openai-compatible/openai-compatible-client.ts]: Addressed. validateResponsesRequest now uses the Zod-parsed value for tool_choice (parsedRequest.tool_choice), so the responses path no longer bypasses parsed output for this field.\n\nI also tightened openAiCompatibleResponsesRequestSchema tool_choice to explicit OpenAI-compatible shapes to preserve type safety end-to-end.\n\nValidation completed with:\n- bun run test -- __tests__/lib/ai-openai-compatible.test.ts\n- bun run test -- __tests__/api/ai/chat-upstream-pipeline-tools.test.ts\n- bun run validate

File: src/lib/ai/openai-compatible/openai-compatible-client.ts [hunk 11]
```diff
@@ -63,6 +74,36 @@ function validateChatRequest(request: OpenAiCompatibleChatCompletionsRequest) {
...
+    parallel_tool_calls: parsedRequest.parallel_tool_calls,
+    reasoning: parsedRequest.reasoning,
+  };
+}
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/lib/ai/openai-compatible/openai-compatible-client.ts]: Fixed in src/lib/ai/openai-compatible/openai-compatible-client.ts by returning parsed tool_choice from validateResponsesRequest (tool_choice now reads from parsedRequest, not raw request). To keep this type-safe, I also tightened the schema in src/types/schemas/ai-openai-compatible.ts so responses tool_choice validates against concrete OpenAI-supported shapes instead of a generic record.\n\nVerification run:\n- bun run test -- __tests__/lib/ai-openai-compatible.test.ts\n- bun run test -- __tests__/api/ai/chat-upstream-pipeline-tools.test.ts\n- bun run validate

File: __tests__/api/ai/chat-upstream-error.test.ts [hunk 12]
```diff
@@ -74,61 +100,78 @@ describe("formatErrorMessage", () => {
...
     expect(result.status).toBe(502);
+    expect(result.kind).toBe("upstream");
   });
 
-  it("maps null/undefined errors to 502", () => {
-    expect(resolveErrorResponse(null).status).toBe(502);
-    expect(resolveErrorResponse(undefined).status).toBe(502);
+  it("maps null/undefined errors to 502 with kind 'upstream'", () => {
+    expect(resolveErrorResponse(null).kind).toBe("upstream");
+    expect(resolveErrorResponse(undefined).kind).toBe("upstream");
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply __tests__/api/ai/chat-upstream-error.test.ts]: # Suggestion Implemented ✅

## Changes Made
- Restored `.status` assertions alongside the new `.kind` assertions in both `maps null/undefined errors to 502` and `maps non-object errors to 502` test cases
- Each edge case now asserts both `status: 502` and `kind: "upstream"`, ensuring neither property can regress undetected

## Validation
- `bun run validate` passed — 0 errors, 0 warnings
- All 27 tests in `chat-upstream-error.test.ts` pass

## Commit
- `e3924620` fix(test): restore status code assertions for edge-case error tests

File: src/types/schemas/ai-openai-compatible.ts [hunk 13]
```diff
@@ -125,6 +142,71 @@ export type OpenAiCompatibleChatCompletionsRequest = z.infer<
...
+  strict: z.boolean().nullable().optional(),
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/types/schemas/ai-openai-compatible.ts]: # Suggestion Analysis

## Decision: Not Implemented

## Reasoning

The `nullable()` on `strict` is intentional for OpenAI Responses API SDK compatibility. The SDK type at `node_modules/openai/resources/responses/responses.d.ts:303` requires `strict: boolean | null` (not optional — it is required and must be `boolean | null`).

The `validateResponsesRequest` function at `openai-compatible-client.ts:89` already normalizes correctly in the **opposite direction** from what is suggested: `tool.strict ?? null` converts `undefined → null` to satisfy the SDK requirement.

Normalizing `null → undefined` as suggested would break the downstream `client.responses.create()` call because the SDK expects `null`, not `undefined`, for this field.

## Evidence
- `node_modules/openai/resources/responses/responses.d.ts:303`: `strict: boolean | null;`
- `openai-compatible-client.ts:89`: `strict: tool.strict ?? null`

File: src/lib/ai/openai-compatible/openai-compatible-client.ts [hunk 14]
```diff
@@ -63,6 +74,36 @@ function validateChatRequest(request: OpenAiCompatibleChatCompletionsRequest) {
...
+    model: parsedRequest.model,
+    input: parsedRequest.input,
+    temperature: parsedRequest.temperature,
+    top_p: parsedRequest.top_p,
+    max_output_tokens: parsedRequest.max_output_tokens,
+    tools: normalizedTools,
+    tool_choice: parsedRequest.tool_choice,
+    parallel_tool_calls: parsedRequest.parallel_tool_calls,
+    reasoning: parsedRequest.reasoning,
+  };
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/lib/ai/openai-compatible/openai-compatible-client.ts]: # Suggestion Implemented

## Changes Made
- Removed 8 redundant property re-assignments (`model`, `input`, `temperature`, `top_p`, `max_output_tokens`, `tool_choice`, `parallel_tool_calls`, `reasoning`) from the `validateResponsesRequest` return statement
- Simplified to `{ ...parsedRequest, tools: normalizedTools }` — only the meaningful override remains

## Validation
- `bun run type-check` passed with 0 errors — the `.loose()` index signature ensures type compatibility with the return type

File: src/lib/ai/openai-compatible/openai-compatible-client.ts [hunk 15]
```diff
@@ -48,10 +58,19 @@ function resolveClient(args: {
...
   if (existingClient) return existingClient;
 
+  let timeout = args.timeoutMs;
+  if (!Number.isFinite(timeout)) {
+    console.warn("[AI] No timeout configured; defaulting to client timeout.", {
+      baseUrl: args.baseUrl,
+      timeoutMs: DEFAULT_TIMEOUT_MS,
+    });
+    timeout = DEFAULT_TIMEOUT_MS;
+  }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/lib/ai/openai-compatible/openai-compatible-client.ts]: # Suggestion Implemented ✅

## Changes Made
- Split the `!Number.isFinite(timeout)` guard into two branches: `undefined` (normal omission) silently defaults to `DEFAULT_TIMEOUT_MS`, while truly abnormal values (`NaN`/`Infinity`) still warn and default.
- This eliminates the log noise on every normal client creation where callers omit the optional `timeoutMs` parameter.

## Validation
- `bun run validate` passed - 0 errors, 0 warnings

## Commit
- `COMMIT_HASH` fix(ai): only warn for abnormal timeout values, not undefined omission
